backend/src/main/java/ai/verta/modeldb/experimentRun/subtypes/PredicatesHandler.java

```diff
@@ -94,6 +94,66 @@
     return InternalFuture.failedStage(new InvalidArgumentException("Predicate cannot be handled"));
   }
 
+  private InternalFuture<QueryFilterContext> processTagsPredicate(
+      long index, KeyValueQuery predicate) {
+    try {
+      final var value = predicate.getValue();
+      var operator = predicate.getOperator();
+      //  We will manage this operator at bottom of method based on `NE` and `NOT_CONTAIN` using
+      // `IN` OR `NOT IN` query
+      if (operator.equals(OperatorEnum.Operator.NOT_CONTAIN)) {
+        operator = OperatorEnum.Operator.CONTAIN;
+      } else if (operator.equals(OperatorEnum.Operator.NE)) {
+        operator = OperatorEnum.Operator.EQ;
+      }
+      final var finalOperator = operator;
+
+      final var valueBindingName = String.format("v_t_%d", index);
+
+      var sql = "select distinct experiment_run_id from tag_mapping where entity_name=:entity_name";
+      sql += " and ";
+
+      final var colValue = "tags";
+      var queryContext =
+          new QueryFilterContext().addBind(q -> q.bind("entity_name", "ExperimentRunEntity"));
+
+      switch (value.getKindCase()) {
+        case STRING_VALUE:
+          sql += applyOperator(finalOperator, colValue, ":" + valueBindingName);
+          queryContext =
+              queryContext.addBind(
+                  q -> q.bind(valueBindingName, wrapValue(finalOperator, value.getStringValue())));
+          break;
+        case LIST_VALUE:
+          List<Object> valueList = new LinkedList<>();
+          for (final var item : value.getListValue().getValuesList()) {
+            if (item.getKindCase().ordinal() == Value.KindCase.STRING_VALUE.ordinal()) {
+              valueList.add(item.getStringValue());
+            }
+          }
+
+          sql += applyOperator(finalOperator, colValue, "<" + valueBindingName + ">");
+          queryContext = queryContext.addBind(q -> q.bindList(valueBindingName, valueList));
+          break;
+        default:
+          return InternalFuture.failedStage(
+              new UnimplementedException("Unknown 'Value' type recognized"));
```

Comment:
It's either unknown or recognized. It can't be both.

Reply:
updated message